Introduction
------------

mapDamage2 is Python/R scripts that tracks and quantify DNA damage patterns among ancient DNA sequencing reads generated by Next-Generation Sequencing platforms. 

### Requirements

* R is mandatory (version >= 2.15.1) and must be present in your $PATH.
* Python version >= 2.6
* pysam, python package, an interface for reading/writing SAM/BAM files

* R libraries:
  -inline
  -gam
  -Rccp
  -RcppGSL
  -ggplot2 (>=0.9.2)

### Installation

mapDamage2 has been successfully tested on GNU/Linux and MacOSX environments.

install the mapDamage package:
download the tarball archive
untar the archive
  tar xfz mapDamage-0.4.tgz
  cd mapDamage-0.4
run the following command if you have administrator rights
  sudo python setup.py install
otherwise install it locally
  python setup.py install --user
Then, $HOME/.local/bin must be in your PATH



To install R, follow the instructions available at
http://www.r-project.org/
Install the missing R packages by running in a R console and selecting a CRAN mirror:
install.packages("inline")
install.packages("gam")
install.packages("Rcpp")
install.packages("ggplot2")

For RcppGSL, download the tarball at the following url:
http://cran.r-project.org/web/packages/RcppGSL/index.html
and install it:
install.packages("RcppGSL_0.2.0.tar.gz")

### Examples and datasets
The original page with examples, datasets and result files is there:

http://geogenetics.ku.dk/publications/mapdamage/

News in version 0.4
---------------------


Usage
-----


Citation
--------
If you use this script, please cite the following publication: Ginolhac A, Rasmussen M, Gilbert MT, Willerslev E, Orlando L.
mapDamage: testing for damage patterns in ancient DNA sequences. Bioinformatics. 2011 27(15):2153-5
http://bioinformatics.oxfordjournals.org/content/27/15/2153

FAQ
-----------

Q: I got this error: could not find function "ggtitle"
A: update your ggplot2 package to at least version 0.9.2 by running
update.packages("ggplot2") and selecting a CRAN mirror

Contact
-------
Please report bugs and suggest possible improvements to Aurelien Ginolhac, Mikkel Schubert or Hákon Jónsson by email. 
