Introduction
------------

mapDamage is a Perl script that tracks DNA damage patterns among ancient DNA sequencing reads generated by Next-Generation Sequencing platforms. 
In addition to the mapDamage Perl script, SAMtools, BEDtools and R are mandatory and must be present in your $PATH.
http://sourceforge.net/projects/samtools/files/samtools/
http://code.google.com/p/bedtools/downloads/list
http://www.r-project.org/

The original page with examples, datasets and result files is there:
http://geogenetics.ku.dk/publications/mapdamage/

News in version 0.3.6
---------------------

* exchange colors for adenine/cytosine in fragmentation patterns, could be confusing with G>A in blue
* bug fix with BAM files obtained from tmap and bowtie
* hard clipping support

Usage
-----

Three main commands can be used: map, merge and plot. The -c option in map allows to chain all the three steps. A detailed description is presented below:

* 1. map

    ./mapDamage-0.3.3.pl map -i input_SAM -d directory -r reference_fasta -c -t my_title

    REQUIRED
        -i : input SAM-file without header. SAM format, version 1.4 is described in this pdf file
        -r : input reference fasta file, with headers identical to the ones in the input_SAM file

    OPTIONAL
        -l : maximal read length, in nucleotides to consider [default: 70]
        -a : size, in nucleotides, of the genomic region to be retrieved before and after reads [default: 10]
        -d : folder for writing output-files, if not already present, this folder will be automatically created [default name: results_input_SAM.filtered]
        -j : second input SAM-file without header; reads present in both SAM files will not be processed
        -u : removes all non-unique reads based on the X1 and XT tags from the BWA mapper
        -k : keep bed file and fasta file containing all the genomic regions [default: delete these files]
        -f : outputs all read aligned against the reference genome, in a fasta file format
        -c : complete analysis, map will be automatically followed by both merge and plot steps [default: not active]

        when -c is active, the following are available to control the plot step:
        -b : the number of reference nucleotides to consider for ploting base composition in the region located upstream and downstream of every read [default: 10]
        -y : graphical y-axis limit for nucleotide misincorporation frequencies [default: 0.30]
        -m : read length, in nucleotides, considered for plotting nucleotide misincorporations [default: 25]

* 2. merge

    ./mapDamage-0.3.3.pl merge -d directory

    REQUIRED
        -d : folder with output-files generated by the map command

        OPTIONAL
        -c : complete analysis, merge will be automatically followed by the plot step [default: not active]

* 3. plot

    ./mapDamage-0.3.3.pl plot -d directory

    REQUIRED
        -d : folder with output-files generated by the merge command

    OPTIONAL
        -l : read length, in nucleotides, considered for plotting nucleotide misincorporations [default: 25]
        -m : graphical y-axis limit for nucleotide misincorporation frequencies [default: 0.30]
        -a : the number of reference nucleotides to consider for ploting base composition in the region located upstream and downstream of every read [default: 10]
        -t : title used for both graph and filename [default: folder name without extension]


Citation
--------
If you use this script, please cite the following publication: Ginolhac A, Rasmussen M, Gilbert MT, Willerslev E, Orlando L.
mapDamage: testing for damage patterns in ancient DNA sequences. Bioinformatics. 2011 27(15):2153-5
http://bioinformatics.oxfordjournals.org/content/27/15/2153

Limitations
-----------
The positions used in fragmentation files are not compatible with read length > 99999 nucleotides. This is compatible with all sequencing technologies currently available.
Contact
-------
Please report bugs and suggest possible improvements to Aurelien Ginolhac by email. 
